### ISSUE TITLE
`replaceState` on scroll after visiting external site

### ISSUE THREAD

**Issue by altano** [vercel/next.js#53347]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
npx --no-install next info

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:23 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6020
    Binaries:
      Node: 20.3.1
      npm: 9.6.7
      Yarn: N/A
      pnpm: 8.6.5
    Relevant Packages:
      next: 13.4.13-canary.6
      eslint-config-next: 13.4.6
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://bright.codehike.org

### To Reproduce

I haven't been able to create a minimal reproduction on Stackblitz yet, but I'm observing this issue on multiple public Next.js sites so I thought I'd report it anyway.

# Example 1 – bright

1. Go to https://bright.codehike.org
2. Click the ["the Next.js docs" link](https://nextjs.org/docs/app/building-your-application/configuring/mdx)
3. Click Back button
4. Scroll around on the page. If nothing happens, mouse over the "recipes" link.
5. Either the scrolling around OR mousing over the recipes link will trigger a navigation back to the "the Next.js docs"
6. Clicking "Back" will NOT take you back to bright.codehike.org, as it appears the navigation was a `replaceState` and your history is now busted.

## Video Demo:
https://github.com/vercel/next.js/assets/1009/09788191-ee28-4123-b1af-d2faa17b260c

# Example 2 – my personal site

1. Go to https://alan.norbauer.com/articles/bookcision
2. Click the ["View this article's source code" link](https://github.com/altano/alan.norbauer.com/tree/main/src/content/articles/bookcision/index.mdx) which links to github
3. Click Back button
4. Scroll around on the page.
5. Scrolling around will trigger a navigation back to github
6. Clicking "Back" will NOT take you back to my site, as it appears the navigation was a `replaceState` and your history is now busted.

## Video Demo:
https://github.com/vercel/next.js/assets/1009/1714b018-fcac-4c53-b077-6f73997ed94e

### Describe the Bug

Here's what I can gather:

1. Navigating to an external site is key to triggering the issue.
2. I can reproduce the issue on desktop Chrome and mobile Safari
3. These are pretty basic Next.js sites. Bright's source is [here](https://github.com/code-hike/bright/tree/bb61dada0f95cd077ebeda37722a52095757f6d0/web) and my site's source is [here](https://github.com/altano/alan.norbauer.com).
4. I _CANNOT_ reproduce this with the devtools open, making debugging extremely difficult.
5. This reproduces in both dev mode and production deploys.
6. [This](https://stackblitz.com/edit/stackblitz-starters-8nfdhk?file=src%2Fapp%2Fpage.tsx) is my unsuccessful attempt to reproduce in stackblitz
7. In mobile Safari I'm more likely to get stuck in a loop where scrolling isn't even required to trigger the erroneous navigation. When I hit back it immediately shoves me forward. I can keep hitting back and my personal site will flash before I'm navigated back to github, stuck there:
https://github.com/vercel/next.js/assets/1009/7f7227eb-ea92-4c84-ae4a-74422a599a6f

### Expected Behavior

I would not expect scrolling to trigger a history `replaceState` or any navigation at all.

### Which browser are you using? (if relevant)

Chrome 115.0.5790.114 (arm64)

<sub>[NEXT-1511](https://linear.app/vercel/issue/NEXT-1511/replacestate-on-scroll-after-visiting-external-site)</sub>

**Comment by steveluscher** [comment]
I was able to get a stack trace by:

1. Loading the site on my phone
2. Performing the repro up until step 3
3. Connecting a debugger to mobile Chrome using `chrome://inspect` from the desktop
4. `window.addEventListener('beforeunload', e => { debugger; })`
5. Resuming the repro steps from step 4

I'll get you an unminified stack trace in a bit.

**Comment by steveluscher** [comment]
It's `location.assign(canonicalUrl)` that causes the errant navigation, here:

https://github.com/vercel/next.js/blob/cdbea3e7da38952a9e706270d8548139bbfe192f/packages/next/src/client/components/app-router.tsx#L362-L383

👋🏻 @sophiebits. RIP #49058.

**Comment by steveluscher** [comment]
Question is, who else from the old crew can we drag into this PR? 🤣

**Comment by ReeceHumphreys** [comment]
Experiencing the same issue. Tried replicating it with the next.js starter repo but was unable to.

**Comment by michel-kraemer** [comment]
I was able to reproduce it as follows:

1. Run `npx create-next-app@latest` to create a new project.
2. Use the default values for everything:
```
✔ What is your project named? … my-app
✔ Would you like to use TypeScript? … Yes
✔ Would you like to use ESLint? … Yes
✔ Would you like to use Tailwind CSS? … Yes
✔ Would you like to use `src/` directory? … No
✔ Would you like to use App Router? (recommended) … Yes
✔ Would you like to customize the default import alias? … No
```
3. Open `next.config.js` and add `output: "export"`:
```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: "export",
}

module.exports = nextConfig
```
4. Replace the contents of `app/page.tsx` with the following code:
```js
import Link from 'next/link'

export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24 m-96">
      <Link href="/">self</Link><br />
      <Link href="https://github.com">GitHub</Link>
    </main>
  )
}
```
5. Run `npm run build` to create a static build
6. Run a web server (e.g. `npx serve -l 3000 out`)
7. Open http://localhost:3000 in your web browser
8. Scroll down and click on the "GitHub" link, then go back, scroll up and hover your mouse cursor over the "self" link (without clicking it!). The page should change to "GitHub" again and the back button will not work correctly any longer as described by @altano.

It seems the "self" link is the culprit. I haven't tested if it also happens if you link to another internal page, but as soon as I remove the self link, everything works OK.

I hope this helps!

**Comment by michel-kraemer** [comment]
I'm surprised not more people are affected by this issue. All I had to do to reproduce it was to create a new project, enable static export, and add an external link. This should basically affect all apps using the app router and external links.

Nevertheless, I created a workaround for the website I'm currently working on. I didn't want to find all external links in my app and replace them with a normal `<a>` tag, so I created a small wrapper around "next/link" instead. It just renders `<a>` for external links and `<Link>` for internal ones.

Click on the following (internal) link if you're interested (pun intended 😉):
https://github.com/steep-wms/steep-wms.github.io/blob/ce98ee8c0bba282ac79d26d7e78c9343cc7ca9b2/src/components/LinkFix.tsx
